File: 05_Open_source_ProjectRepository/AIAgentFramework/EnlightenmentLighthouseRuntime/elr/token_manager.go
// Package elr implements token management for Enlightenment Lighthouse Runtime
package elr

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/google/uuid"
)

// TokenStatus represents the status of a token
type TokenStatus string

const (
	// TokenStatusActive represents an active token
	TokenStatusActive TokenStatus = "active"
	// TokenStatusRevoked represents a revoked token
	TokenStatusRevoked TokenStatus = "revoked"
	// TokenStatusExpired represents an expired token
	TokenStatusExpired TokenStatus = "expired"
)

// Token represents a token instance
type Token struct {
	ID          string       `json:"id"`
	Secret      string       `json:"secret"`
	Description string       `json:"description"`
	CreatedAt   time.Time    `json:"created_at"`
	ExpiresAt   time.Time    `json:"expires_at"`
	Status      TokenStatus  `json:"status"`
}

// TokenManager represents the token manager
type TokenManager struct {
	tokenFile string
	tokens    []Token
}

// NewTokenManager creates a new token manager
func NewTokenManager(tokenFile string) *TokenManager {
	return &TokenManager{
		tokenFile: tokenFile,
		tokens:    []Token{},
	}
}

// LoadTokens loads tokens from file
func (tm *TokenManager) LoadTokens() error {
	if _, err := os.Stat(tm.tokenFile); os.IsNotExist(err) {
		// Token file doesn't exist, initialize with empty tokens
		tm.tokens = []Token{}
		return nil
	}

	data, err := os.ReadFile(tm.tokenFile)
	if err != nil {
		return fmt.Errorf("failed to read token file: %w", err)
	}

	var tokenData struct {
		Tokens       []Token `json:"tokens"`
		LastUpdated  int64   `json:"last_updated"`
	}

	if err := json.Unmarshal(data, &tokenData); err != nil {
		return fmt.Errorf("failed to unmarshal token data: %w", err)
	}

	tm.tokens = tokenData.Tokens
	return nil
}

// SaveTokens saves tokens to file
func (tm *TokenManager) SaveTokens() error {
	tokenData := struct {
		Tokens      []Token `json:"tokens"`
		LastUpdated int64   `json:"last_updated"`
	}{
		Tokens:      tm.tokens,
		LastUpdated: time.Now().Unix(),
	}

	data, err := json.MarshalIndent(tokenData, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal token data: %w", err)
	}

	// Create directory if not exists
	if err := os.MkdirAll(filepath.Dir(tm.tokenFile), 0755); err != nil {
		return fmt.Errorf("failed to create token directory: %w", err)
	}

	if err := os.WriteFile(tm.tokenFile, data, 0644); err != nil {
		return fmt.Errorf("failed to write token file: %w", err)
	}

	return nil
}

// GenerateToken generates a new token
func (tm *TokenManager) GenerateToken(description string) (string, error) {
	tokenID := uuid.New().String()
	tokenSecret := fmt.Sprintf("%x", sha256.Sum256([]byte(uuid.New().String())))

	token := Token{
		ID:          tokenID,
		Secret:      tokenSecret,
		Description: description,
		CreatedAt:   time.Now(),
		ExpiresAt:   time.Now().Add(7 * 24 * time.Hour), // 7 days expiration
		Status:      TokenStatusActive,
	}

	tm.tokens = append(tm.tokens, token)

	if err := tm.SaveTokens(); err != nil {
		return "", fmt.Errorf("failed to save token: %w", err)
	}

	return fmt.Sprintf("%s.%s", tokenID, tokenSecret), nil
}

// ValidateToken validates a token
func (tm *TokenManager) ValidateToken(tokenString string) (bool, string) {
	if tokenString == "" {
		return false, "Token is required"
	}

	// Parse token string
	parts := splitToken(tokenString)
	if len(parts) != 2 {
		return false, "Invalid token format"
	}

	tokenID, tokenSecret := parts[0], parts[1]

	// Find token
	for _, token := range tm.tokens {
		if token.ID == tokenID && token.Secret == tokenSecret {
			// Check if token is expired
			if time.Now().After(token.ExpiresAt) {
				return false, "Token has expired"
			}

			// Check if token is active
			if token.Status != TokenStatusActive {
				return false, "Token is not active"
			}

			return true, "Token is valid"
		}
	}

	return false, "Token not found"
}

// RefreshToken refreshes a token
func (tm *TokenManager) RefreshToken(oldTokenString, description string) (string, error) {
	// Validate old token
	valid, message := tm.ValidateToken(oldTokenString)
	if !valid {
		return "", fmt.Errorf("invalid old token: %s", message)
	}

	// Parse old token
	parts := splitToken(oldTokenString)
	tokenID := parts[0]

	// Revoke old token
	for i, token := range tm.tokens {
		if token.ID == tokenID {
			tm.tokens[i].Status = TokenStatusRevoked
			break
		}
	}

	// Generate new token
	newToken, err := tm.GenerateToken(description)
	if err != nil {
		return "", fmt.Errorf("failed to generate new token: %w", err)
	}

	return newToken, nil
}

// ListTokens lists all tokens
func (tm *TokenManager) ListTokens() []map[string]interface{} {
	result := make([]map[string]interface{}, 0, len(tm.tokens))

	for _, token := range tm.tokens {
		tokenInfo := map[string]interface{}{
			"id":          token.ID,
			"description": token.Description,
			"created_at":  token.CreatedAt.Unix(),
			"expires_at":  token.ExpiresAt.Unix(),
			"status":      token.Status,
			"expired":     time.Now().After(token.ExpiresAt),
		}
		result = append(result, tokenInfo)
	}

	return result
}

// RevokeToken revokes a token
func (tm *TokenManager) RevokeToken(tokenID string) error {
	for i, token := range tm.tokens {
		if token.ID == tokenID {
			tm.tokens[i].Status = TokenStatusRevoked
			if err := tm.SaveTokens(); err != nil {
				return fmt.Errorf("failed to save token: %w", err)
			}
			return nil
		}
	}

	return fmt.Errorf("token not found")
}

// splitToken splits a token string into ID and secret
func splitToken(token string) []string {
	parts := make([]string, 0, 2)
	for i, c := range token {
		if c == '.' {
			parts = append(parts, token[:i], token[i+1:])
			break
		}
	}
	return parts
}

File: 05_Open_source_ProjectRepository/AIAgentFramework/EnlightenmentLighthouseRuntime/elr/token_manager.py
#!/usr/bin/env python3
"""
ELR容器令牌管理器
负责生成、验证和管理ELR容器的访问令牌
"""

import json
import time
import os
import hashlib
import uuid

class TokenManager:
    """令牌管理器"""
    
    def __init__(self, token_file="elr_token.json"):
        """初始化令牌管理器"""
        self.token_file = token_file
        self.token_data = self.load_tokens()
    
    def load_tokens(self):
        """从文件加载令牌数据"""
        if os.path.exists(self.token_file):
            try:
                with open(self.token_file, 'r', encoding='utf-8') as f:
                    return json.load(f)
            except Exception as e:
                print(f"Error loading token file: {e}")
        return {
            "tokens": [],
            "last_updated": time.time()
        }
    
    def save_tokens(self):
        """保存令牌数据到文件"""
        try:
            self.token_data["last_updated"] = time.time()
            with open(self.token_file, 'w', encoding='utf-8') as f:
                json.dump(self.token_data, f, indent=2, ensure_ascii=False)
            return True
        except Exception as e:
            print(f"Error saving token file: {e}")
            return False
    
    def generate_token(self, description="ELR Container Token"):
        """生成新令牌"""
        token_id = str(uuid.uuid4())
        token_secret = hashlib.sha256(str(uuid.uuid4()).encode()).hexdigest()
        token = {
            "id": token_id,
            "secret": token_secret,
            "description": description,
            "created_at": time.time(),
            "expires_at": time.time() + (7 * 24 * 3600),  # 7天过期
            "status": "active"
        }
        
        self.token_data["tokens"].append(token)
        self.save_tokens()
        
        # 返回完整令牌（id + secret）
        return f"{token_id}.{token_secret}"
    
    def validate_token(self, token):
        """验证令牌"""
        if not token:
            return False, "Token is required"
        
        try:
            token_parts = token.split('.')
            if len(token_parts) != 2:
                return False, "Invalid token format"
            
            token_id, token_secret = token_parts
            
            for t in self.token_data["tokens"]:
                if t["id"] == token_id and t["secret"] == token_secret:
                    # 检查令牌是否过期
                    if time.time() > t["expires_at"]:
                        return False, "Token has expired"
                    # 检查令牌状态
                    if t["status"] != "active":
                        return False, "Token is not active"
                    return True, "Token is valid"
            
            return False, "Token not found"
        except Exception as e:
            return False, f"Error validating token: {e}"
    
    def refresh_token(self, old_token, description="Refreshed ELR Container Token"):
        """刷新令牌"""
        # 验证旧令牌
        valid, message = self.validate_token(old_token)
        if not valid:
            return None, message
        
        # 禁用旧令牌
        token_parts = old_token.split('.')
        token_id = token_parts[0]
        
        for i, t in enumerate(self.token_data["tokens"]):
            if t["id"] == token_id:
                self.token_data["tokens"][i]["status"] = "revoked"
                break
        
        # 生成新令牌
        new_token = self.generate_token(description)
        return new_token, "Token refreshed successfully"
    
    def list_tokens(self):
        """列出所有令牌"""
        tokens = []
        for t in self.token_data["tokens"]:
            token_info = {
                "id": t["id"],
                "description": t["description"],
                "created_at": t["created_at"],
                "expires_at": t["expires_at"],
                "status": t["status"],
                "expired": time.time() > t["expires_at"]
            }
            tokens.append(token_info)
        return tokens
    
    def revoke_token(self, token_id):
        """撤销令牌"""
        for i, t in enumerate(self.token_data["tokens"]):
            if t["id"] == token_id:
                self.token_data["tokens"][i]["status"] = "revoked"
                self.save_tokens()
                return True, "Token revoked successfully"
        return False, "Token not found"

if __name__ == "__main__":
    # 测试令牌管理器
    tm = TokenManager()
    
    # 生成新令牌
    token = tm.generate_token("Test Token")
    print(f"Generated token: {token}")
    
    # 验证令牌
    valid, message = tm.validate_token(token)
    print(f"Token validation: {valid}, {message}")
    
    # 列出令牌
    tokens = tm.list_tokens()
    print(f"Tokens: {json.dumps(tokens, indent=2)}")
    
    # 刷新令牌
    new_token, message = tm.refresh_token(token, "Refreshed Test Token")
    print(f"Refreshed token: {new_token}, {message}")
    
    # 验证新令牌
    valid, message = tm.validate_token(new_token)
    print(f"New token validation: {valid}, {message}")
    
    # 验证旧令牌
    valid, message = tm.validate_token(token)
    print(f"Old token validation: {valid}, {message}")
